File: .js
// Frontend (React) - App.js
import React, { useState, useEffect } from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import Dashboard from './components/Dashboard';
import Users from './components/Users';
import Products from './components/Products';
import Orders from './components/Orders';
import Sidebar from './components/Sidebar';
import './styles/Admin.css';

function App() {
  return (
    <BrowserRouter>
      <div className="admin-container">
        <Sidebar />
        <Routes>
          <Route path="/" element={<Dashboard />} />
          <Route path="/users" element={<Users />} />
          <Route path="/products" element={<Products />} />
          <Route path="/orders" element={<Orders />} />
        </Routes>
      </div>
    </BrowserRouter>
  );
}

// Backend (Node.js/Express)
const express = require('express');
const mongoose = require('mongoose');
const cors = require('cors');

const app = express();

// Middleware
app.use(cors());
app.use(express.json());

// MongoDB connection
mongoose.connect('mongodb://localhost/admin-panel', {
  useNewUrlParser: true,
  useUnifiedTopology: true
});

// User Model
const User = mongoose.model('User', {
  name: String,
  email: String,
  role: String,
  status: String,
  createdAt: { type: Date, default: Date.now }
});

// Product Model
const Product = mongoose.model('Product', {
  name: String,
  price: Number,
  stock: Number,
  category: String,
  createdAt: { type: Date, default: Date.now }
});

// Order Model
const Order = mongoose.model('Order', {
  userId: mongoose.Schema.Types.ObjectId,
  products: [{
    productId: mongoose.Schema.Types.ObjectId,
    quantity: Number,
    price: Number
  }],
  total: Number,
  status: String,
  createdAt: { type: Date, default: Date.now }
});

// Routes
app.get('/api/stats', async (req, res) => {
  try {
    const stats = {
      users: await User.countDocuments(),
      products: await Product.countDocuments(),
      orders: await Order.countDocuments(),
      revenue: await Order.aggregate([
        { $match: { status: 'completed' } },
        { $group: { _id: null, total: { $sum: '$total' } } }
      ])
    };
    res.json(stats);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

app.get('/api/users', async (req, res) => {
  try {
    const users = await User.find().sort({ createdAt: -1 });
    res.json(users);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

app.post('/api/users', async (req, res) => {
  try {
    const user = new User(req.body);
    await user.save();
    res.status(201).json(user);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// React Components
// components/Dashboard.js
function Dashboard() {
  const [stats, setStats] = useState(null);

  useEffect(() => {
    fetch('http://localhost:5000/api/stats')
      .then(res => res.json())
      .then(data => setStats(data));
  }, []);

  return (
    <div className="dashboard">
      {stats && (
        <div className="stats-grid">
          <div className="stat-card">
            <h3>Total Users</h3>
            <div className="number">{stats.users}</div>
          </div>
          {/* More stat cards */}
        </div>
      )}
    </div>
  );
}

// components/Users.js
function Users() {
  const [users, setUsers] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    fetchUsers();
  }, []);

  const fetchUsers = async () => {
    try {
      const response = await fetch('http://localhost:5000/api/users');
      const data = await response.json();
      setUsers(data);
      setLoading(false);
    } catch (error) {
      console.error('Error fetching users:', error);
      setLoading(false);
    }
  };

  return (
    <div className="users-page">
      <h2>Users Management</h2>
      {loading ? (
        <div>Loading...</div>
      ) : (
        <table className="data-table">
          <thead>
            <tr>
              <th>Name</th>
              <th>Email</th>
              <th>Role</th>
              <th>Status</th>
              <th>Actions</th>
            </tr>
          </thead>
          <tbody>
            {users.map(user => (
              <tr key={user._id}>
                <td>{user.name}</td>
                <td>{user.email}</td>
                <td>{user.role}</td>
                <td>{user.status}</td>
                <td>
                  <button onClick={() => handleEdit(user._id)}>Edit</button>
                  <button onClick={() => handleDelete(user._id)}>Delete</button>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </div>
  );
}

File: .py
# app.py
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy
from flask_cors import CORS
from werkzeug.security import generate_password_hash, check_password_hash

app = Flask(__name__)
CORS(app)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///users.db'
db = SQLAlchemy(app)

class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(80), nullable=False)
    email = db.Column(db.String(120), unique=True, nullable=False)
    password = db.Column(db.String(120), nullable=False)

@app.route('/api/signup', methods=['POST'])
def signup():
    data = request.json
    if User.query.filter_by(email=data['email']).first():
        return jsonify({'error': 'Email already exists'}), 400
    
    hashed_password = generate_password_hash(data['password'])
    new_user = User(
        name=data['name'],
        email=data['email'],
        password=hashed_password
    )
    db.session.add(new_user)
    db.session.commit()
    return jsonify({'message': 'User created successfully'}), 201

@app.route('/api/login', methods=['POST'])
def login():
    data = request.json
    user = User.query.filter_by(email=data['email']).first()
    
    if user and check_password_hash(user.password, data['password']):
        return jsonify({
            'message': 'Login successful',
            'user': {
                'name': user.name,
                'email': user.email
            }
        })
    return jsonify({'error': 'Invalid credentials'}), 401

if __name__ == '__main__':
    db.create_all()
    app.run(debug=True)






























# app.py
from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy
from flask_cors import CORS
from datetime import datetime
import jwt
import os

app = Flask(__name__)
CORS(app)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///marketplace.db'
app.config['SECRET_KEY'] = 'your-secret-key'
db = SQLAlchemy(app)

# Database Models
class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(80), nullable=False)
    email = db.Column(db.String(120), unique=True, nullable=False)
    password = db.Column(db.String(120), nullable=False)
    is_seller = db.Column(db.Boolean, default=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    products = db.relationship('Product', backref='seller', lazy=True)

class Product(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)
    description = db.Column(db.Text)
    price = db.Column(db.Float, nullable=False)
    image_url = db.Column(db.String(200))
    stock = db.Column(db.Integer, default=0)
    seller_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)

class Cart(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'), nullable=False)
    product_id = db.Column(db.Integer, db.ForeignKey('product.id'), nullable=False)
    quantity = db.Column(db.Integer, default=1)

# API Routes
@app.route('/api/products', methods=['GET', 'POST'])
def products():
    if request.method == 'GET':
        products = Product.query.all()
        return jsonify([{
            'id': p.id,
            'name': p.name,
            'price': p.price,
            'description': p.description,
            'image_url': p.image_url,
            'stock': p.stock,
            'seller': p.seller.name
        } for p in products])
    
    if request.method == 'POST':
        data = request.json
        new_product = Product(
            name=data['name'],
            description=data['description'],
            price=data['price'],
            image_url=data['image_url'],
            stock=data['stock'],
            seller_id=data['seller_id']
        )
        db.session.add(new_product)
        db.session.commit()
        return jsonify({'message': 'Product added successfully'}), 201

@app.route('/api/cart', methods=['GET', 'POST', 'DELETE'])
def cart():
    if request.method == 'GET':
        user_id = request.args.get('user_id')
        cart_items = Cart.query.filter_by(user_id=user_id).all()
        return jsonify([{
            'id': item.id,
            'product': Product.query.get(item.product_id).name,
            'quantity': item.quantity,
            'price': Product.query.get(item.product_id).price
        } for item in cart_items])

if __name__ == '__main__':
    db.create_all()
    app.run(debug=True)
















# app.py
import stripe
from flask import Flask, jsonify, request

stripe.api_key = 'your_stripe_secret_key'

# Order Model
class Order(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'))
    total = db.Column(db.Float, nullable=False)
    status = db.Column(db.String(20), default='pending')
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    payment_intent = db.Column(db.String(200))

# Payment Routes
@app.route('/api/create-payment-intent', methods=['POST'])
def create_payment():
    try:
        data = request.json
        intent = stripe.PaymentIntent.create(
            amount=int(data['amount'] * 100),  # Convert to cents
            currency='usd'
        )
        
        # Create order in database
        order = Order(
            user_id=data['user_id'],
            total=data['amount'],
            payment_intent=intent.id
        )
        db.session.add(order)
        db.session.commit()

        return jsonify({
            'clientSecret': intent.client_secret,
            'order_id': order.id
        })

    except Exception as e:
        return jsonify({'error': str(e)}), 400

@app.route('/api/payment-success', methods=['POST'])
def payment_success():
    data = request.json
    order = Order.query.filter_by(payment_intent=data['payment_intent']).first()
    if order:
        order.status = 'completed'
        db.session.commit()
        return jsonify({'message': 'Payment successful'})
    return jsonify({'error': 'Order not found'}), 404

@app.route('/api/orders/<user_id>', methods=['GET'])
def get_orders(user_id):
    orders = Order.query.filter_by(user_id=user_id).all()
    return jsonify([{
        'id': order.id,
        'total': order.total,
        'status': order.status,
        'date': order.created_at.strftime('%Y-%m-%d %H:%M:%S')
    } for order in orders])




























# app.py - Add these new routes and models

class Profile(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('user.id'))
    avatar_url = db.Column(db.String(200))
    bio = db.Column(db.Text)
    phone = db.Column(db.String(20))
    address = db.Column(db.String(200))
    store_name = db.Column(db.String(100))
    is_seller = db.Column(db.Boolean, default=False)

class SellerStats(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    seller_id = db.Column(db.Integer, db.ForeignKey('user.id'))
    total_sales = db.Column(db.Float, default=0)
    total_orders = db.Column(db.Integer, default=0)
    rating = db.Column(db.Float, default=0)
    products_count = db.Column(db.Integer, default=0)

@app.route('/api/dashboard/<int:user_id>', methods=['GET'])
def get_dashboard(user_id):
    user = User.query.get_or_404(user_id)
    profile = Profile.query.filter_by(user_id=user_id).first()
    
    if profile.is_seller:
        stats = SellerStats.query.filter_by(seller_id=user_id).first()
        recent_orders = Order.query.filter_by(seller_id=user_id)\
                                 .order_by(Order.created_at.desc())\
                                 .limit(5).all()
        return jsonify({
            'profile': {
                'name': user.name,
                'email': user.email,
                'store_name': profile.store_name,
                'avatar_url': profile.avatar_url,
                'bio': profile.bio
            },
            'stats': {
                'total_sales': stats.total_sales,
                'total_orders': stats.total_orders,
                'rating': stats.rating,
                'products_count': stats.products_count
            },
            'recent_orders': [{
                'id': order.id,
                'amount': order.total,
                'status': order.status,
                'date': order.created_at.strftime('%Y-%m-%d %H:%M:%S')
            } for order in recent_orders]
        })
    else:
        recent_orders = Order.query.filter_by(user_id=user_id)\
                                 .order_by(Order.created_at.desc())\
                                 .limit(5).all()
        return jsonify({
            'profile': {
                'name': user.name,
                'email': user.email,
                'avatar_url': profile.avatar_url
            },
            'recent_orders': [{
                'id': order.id,
                'amount': order.total,
                'status': order.status,
                'date': order.created_at.strftime('%Y-%m-%d %H:%M:%S')
            } for order in recent_orders]
        })

@app.route('/api/profile/update', methods=['POST'])
def update_profile():
    data = request.json
    profile = Profile.query.filter_by(user_id=data['user_id']).first()
    
    for key, value in data.items():
        if hasattr(profile, key):
            setattr(profile, key, value)
    
    db.session.commit()
    return jsonify({'message': 'Profile updated successfully'})
